---
title: "hgcA_processing"
author: "Benjamin D. Peterson"
date: "2023-04-05"
output: html_document
---

```{r setup, include = FALSE}
rm(list = ls())
#setwd("/Users/benjaminpeterson/Documents/research/BLiMMP")
library(tidyverse)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(echo = TRUE)
new.wd <- gsub("\\/[[:alpha:]]+\\/[[:alpha:]]+$", "", getwd())
knitr::opts_knit$set(root.dir = new.wd)
rm(new.wd)
```

```{r read_in_faa, include = FALSE }
library(Biostrings)
hgcA.faa <- readAAStringSet("dataEdited/ABA/hgcA/hgcA.faa")
# Store the HMM score in the metadata
metadata(hgcA.faa)$HMM_scores <- strsplit(hgcA.faa@ranges@NAMES, " ") %>% sapply("[", 2) %>% data.frame()
names(hgcA.faa) <- strsplit(names(hgcA.faa), " ") %>% sapply("[", 1)
```

I identified *hgcA* genes in the assemblies using my GID workflow from HomeBio. I loaded the alignment into Geneious and manually inspected the alignment to remove sequences that hit our exclusion criteria, which are as follows:

1. Missing the canonical cap helix domain
2. Truncated at the C-terminus end and have fewer than 250 amino acids in the alignment. Essentially for this criteria we want to ensure that the transmembrane domains at the N-terminus are represented. 

While the excluded sequences are likely incompletely assembled *hgcA* sequences, we prefer to err on the side of caution when it comes to identifying *hgcA*. A preview of the notes on the sequences we excluded are below:

```{r hgcA_excluded_seqs, echo = FALSE}
excluded.seqs <- readxl::read_xlsx("dataEdited/ABA/hgcA/hgcA_hits_to_exclude.xlsx")
excluded.seqs

```

In the interest of including all the information we have, I generated a dataframe that includes all of the identified sequences with a column denoting the ones that passed criteria and are included in the downstream analyses. The "verified_hgcA" column will be a TRUE/FALSE column, with TRUE marking the sequences that pass the criteria. I included the length of the sequences here as well.


```{r hgcA_ID_notes, echo = 1, message = FALSE}
hgcA.data <- data.frame(seqID = names(hgcA.faa),
                        HMM_score = metadata(hgcA.faa)$HMM_scores,
                        length = width(hgcA.faa)) %>%
  left_join(excluded.seqs) %>%
  mutate(verified_hgcA = is.na(reason_for_exclusion))
rm(excluded.seqs)
```


Next I added in the clustering information and scaffold names. It looks like CD-HIT automatically assigns one of the longest sequences in the cluster as the cluster representative, which is fine with me, so that's what I used for the phylogenetic analysis. 

```{r hgcA_cluster_scaffold, echo = -4, message = FALSE}
hgcA.clstr.data <- read.table("dataEdited/ABA/hgcA/hgcA_cluster_data.tsv",
                              header = TRUE) %>%
  dplyr::rename(seqID = id) %>%
  mutate(cluster_ID = paste("BLI_hgcA_clstr_", stringr::str_pad(clstr, width = 4, side = "left", pad = 0),
                            sep = "")) %>%
  select(seqID, cluster_ID, clstr_rep)
hgcA.scaf.assembly <- read.table("dataEdited/ABA/hgcA/hgcA_G2A.tsv",
                                 col.names = c("seqID", "assemblyID")) %>%
  mutate(scaffoldID = paste(strsplit(seqID, "_") %>% sapply("[", 1),
                            strsplit(seqID, "_") %>% sapply("[", 2),
                            strsplit(seqID, "_") %>% sapply("[", 3),
                            sep = "_"))

hgcA.data <- hgcA.data %>%
  left_join(hgcA.clstr.data) %>%
  left_join(hgcA.scaf.assembly)
rm(hgcA.clstr.data, hgcA.scaf.assembly)
```

----

# Add taxonomy and metabolic assignment data from bins

This is a bit tricky for several reasons:
- The assembly *hgcA* sequences were clustered and one cluster representative was chosen for phylogenetic analysis. This representative does not necessarily match up with what was binned.
- The ORF names won't necessarily match up between bins and assemblies, since Prodigal was called separately on both and used different modes.
- There are multiple bins correspondin to some assembly-derived *hgcA* genes/clusters.

What I'll do is assign each bin an *hgcA* cluster ID, then dereplicate the bins

```{r add_bin_data, echo = -4, message = FALSE}
bin_data <- readRDS('dataEdited/bin_based_analyses/bin_data_aggregate_2_metabolism.rds') %>%
  dplyr::rename(seqID = hgca_id) %>%
  mutate(seqID = strsplit(seqID, ",")) %>%
  unnest(seqID) %>%
  mutate(join_column = paste(seqID %>% strsplit("_") %>% sapply("[", 1),
                             seqID %>% strsplit("_") %>% sapply("[", 2),
                             seqID %>% strsplit("_") %>% sapply("[", 3),
                             sep = "_")) %>%
  select(join_column, hms_id, gtdb_tax, meta_code) %>%
  arrange(join_column, hms_id) %>%
  filter(!duplicated(join_column)) %>%
  left_join(hgcA.data %>%
              select(seqID, cluster_ID) %>%
              mutate(join_column = paste(seqID %>% strsplit("_") %>% sapply("[", 1),
                             seqID %>% strsplit("_") %>% sapply("[", 2),
                             seqID %>% strsplit("_") %>% sapply("[", 3),
                             sep = "_")),
            by = 'join_column') %>%
  select(-join_column) %>%
  group_by(cluster_ID) %>%
  summarize(hms_id = names(which.max(table(hms_id))),
            gtdb_tax = names(which.max(table(gtdb_tax))),
            meta_code = names(which.max(table(meta_code))))
# hgcA.data <- hgcA.data %>%
#   left_join(bin_data, by = "seqID")

```

----

# *hgcA* phylogenetic tree to estimate taxonomic identity and metabolic capacity of each organism carrying *hgcA*

Although we'll average across the cluster for depth calculations, we will want to identify one sequences from each cluster that can be used for phylogenetic reconstruction. Thus, we'll save out a list with that dereplicated list to `dataEdited/ABA/hgcA/hgcA_for_phylogeny.faa`.

```{r save out hgcA_for_phylogeny, message = FALSE}
hgcA.seqs.for.phylogeny <- hgcA.data %>%
  filter(verified_hgcA,
         clstr_rep == 1) %>%
  select(seqID) %>%
  unlist(use.names = FALSE)
hgcA.for.phylogeny.faa <- hgcA.faa[hgcA.seqs.for.phylogeny]
writeXStringSet(hgcA.for.phylogeny.faa,
                filepath = "dataEdited/ABA/hgcA/hgcA_for_phylogeny.faa",
                format = "fasta",
                width = 6000)
```

For references, I took the Hg-MATE database (version 1) and clustered the sequences at 70% identity. I then aligned the output from this to hgcA_for_phylogeny.faa using Muscle (v5.1). I trimmed that alignment to eliminate residues with gaps in >50% of the sequences using trimal (v1.4.rev15). Finally, I generated a rough phylogenetic tree from this alignment using FastTree (v2.1.11). First I looked at the unrooted tree (not shown in this document) to identify the paralog (paralog_Candidatus_Omnitrophica_bacterium_CG1_02_41_171) for rooting purposes.

```{r unrooted_hgcA_phylogeny, include = FALSE}
library(ggtree)
library(treeio)
hgcA.tree.unrooted <- read.newick("dataEdited/ABA/hgcA/hgcA_phylogeny.tree")
pdf("working_directory/BLI_hgcA_tree_unrooted.pdf",
    height = 90,
    width = 8)
ggtree(hgcA.tree.unrooted) +
  geom_tiplab() +
  geom_text2(aes(subset = !isTip, label = node),
             size = 1.5) +
  geom_treescale(x = 9)
dev.off()
```

The branch leading to the paralog in the unrooted tree is #419, so I used that to root the tree. I then saved out a PDF of the rooted tree to the working directory and used that to assign taxonomy to each of the *hgcA* sequences.

```{r rooted_hgcA_phylogeny, message = FALSE, fig.height = 25}
hgcA.tree <- root(hgcA.tree.unrooted,outgroup = "paralog_Candidatus_Omnitrophica_bacterium_CG1_02_41_171",
                  edgelabel = TRUE)
this.study.indices <- which(hgcA.tree$tip.label %in% names(hgcA.for.phylogeny.faa))

# Set up color vector
color.vector <- rep("black", length(hgcA.tree$tip.label))
color.vector[this.study.indices] <- "red"

# Set up vector for names of hgcA seqs that were binned.
hgcA_tree_bin_data <- hgcA.data %>%
  filter(seqID %in% names(hgcA.for.phylogeny.faa)) %>%
  select(seqID, cluster_ID) %>%
  left_join(bin_data) %>%
  filter(!is.na(meta_code)) %>%
  mutate(phyla_fam = paste(gtdb_tax %>% strsplit("p__") %>% sapply("[", 2) %>% strsplit(";c__") %>% sapply("[", 1),
                           gtdb_tax %>% strsplit("c__") %>% sapply("[", 2) %>% strsplit(";o__") %>% sapply("[", 1),
                           gtdb_tax %>% strsplit("o__") %>% sapply("[", 2) %>% strsplit(";f__") %>% sapply("[", 1),
                           gtdb_tax %>% strsplit("f__") %>% sapply("[", 2) %>% strsplit(";g__") %>% sapply("[", 1),
                           sep = ";"))
hgcA_tree_tip_renaming <- paste(hgcA_tree_bin_data$seqID, ":",
                                hgcA_tree_bin_data$hms_id, "-",
                                hgcA_tree_bin_data$phyla_fam, "-",
                                hgcA_tree_bin_data$meta_code,
                                sep = "") %>%
  gsub(";-", "-", .) %>%
  gsub(";-", "-", .)
names(hgcA_tree_tip_renaming) <- hgcA_tree_bin_data$seqID
binned_index <- which(hgcA.tree$tip.label %in% names(hgcA_tree_tip_renaming))
hgcA.tree$tip.label[binned_index] <- hgcA_tree_tip_renaming[hgcA.tree$tip.label[binned_index]]

hgcA.tree.image <- ggtree(hgcA.tree) + 
  geom_tiplab(size = 2.5,
              color = color.vector) +
  geom_treescale(x = 9) +
  geom_text2(aes(subset = !isTip,
                 label = node),
             size = 1.5)
hgcA.tree.image
pdf("working_directory/BLI_hgcA_tree_rooted.pdf",
    height = 50,
    width = 8)
hgcA.tree.image
dev.off()

saveRDS(object = hgcA.tree,
        file = "manuscript/writing/Suppl_Data/hgcA_tree.rds")

```

Next I'll work my way down the tree, assigning metabolic capacity and taxonomic identity to each sequence.

```{r tax_meta_assignment_by_phylogeny, include = FALSE}
library(tidytree)
taxonomic_grouping_hgcA <- vector()
metabolic_grouping_hgcA <- vector()

taxonomic_grouping_hgcA['BLI21_coassembly_000000219078_2'] <- "Firmicutes"
metabolic_grouping_hgcA['BLI21_coassembly_000000219078_2'] <- "FERM"

# This sequence was likely a Spirochaetes sequence, but we'll just call them PVC.
taxonomic_grouping_hgcA['BLI21_coassembly_000000277987_1'] <- "PVC"
metabolic_grouping_hgcA['BLI21_coassembly_000000277987_1'] <- "UNK"

# Due to discrepancy between FERM and RESP annotations in main Kiritimatiellae, we will call all these KIR for metabolism.
taxonomic_grouping_hgcA[c("BLI20_coassembly_000000214677_1", "BLI21_coassembly_000000397409_2")] <- "Kiritimatiellae"
metabolic_grouping_hgcA[c("BLI20_coassembly_000000214677_1", "BLI21_coassembly_000000397409_2")] <- "KIR"

taxonomic_grouping_hgcA['BLI21_coassembly_000000003920_15'] <- "Firmicutes"
metabolic_grouping_hgcA['BLI21_coassembly_000000003920_15'] <- "FERM"

taxonomic_grouping_hgcA[c("BLI21_assembly104_000000083274_2", "BLI21_coassembly_000000136593_2")] <- "Actinobacteria"
metabolic_grouping_hgcA[c("BLI21_assembly104_000000083274_2", "BLI21_coassembly_000000136593_2")] <- "RESP"

# Assigned FERM based on prevalence of obligatory fermentation in Firmicutes in this system. 
taxonomic_grouping_hgcA['BLI21_coassembly_000000858312_1'] <- "Firmicutes"
metabolic_grouping_hgcA['BLI21_coassembly_000000858312_1'] <- "FERM"

# Assigned to RESP based on similarity to an hgcA gene from Peterson et al, 2020 EST paper, which was done in the same system.
taxonomic_grouping_hgcA['BLI21_coassembly_000000724865_2'] <- "Planctomycetes"
metabolic_grouping_hgcA['BLI21_coassembly_000000724865_2'] <- "RESP"

# Gene is binned. Due to discrepancy between FERM and RESP annotations in main Kiritimatiellae, we will call all these KIR for metabolism.
taxonomic_grouping_hgcA['BLI21_coassembly_000000061249_5'] <- "Kiritimatiellae"
metabolic_grouping_hgcA['BLI21_coassembly_000000061249_5'] <- "KIR"

# One of these genes is binned. Both are closed related to gene from Peterson et al, 2020, which has same taxonomic ID and metabolic capacity.
taxonomic_grouping_hgcA[c("BLI21_coassembly_000000018594_7", "BLI21_assembly105_000000164796_1")] <- "Actinobacteria"
metabolic_grouping_hgcA[c("BLI21_coassembly_000000018594_7", "BLI21_assembly105_000000164796_1")] <- "FERM"

# No close genes from this system. Safely within Bacteroidetes. Didn't feel comfortable assigning metabolic capacity though. 
taxonomic_grouping_hgcA["BLI21_assembly102_000000112326_2"] <- "Bacteroidetes"
metabolic_grouping_hgcA["BLI21_assembly102_000000112326_2"] <- "UNK"

# No close genes from this system. Safely within Geobacteraceae. These are always respiratory. 
taxonomic_grouping_hgcA["BLI21_coassembly_000000229519_3"] <- "Geobacteraceae"
metabolic_grouping_hgcA["BLI21_coassembly_000000229519_3"] <- "RESP"

# No close genes from this system. Closest to some Chloroflexi.
taxonomic_grouping_hgcA["BLI21_coassembly_000001026396_2"] <- "Chloroflexi"
metabolic_grouping_hgcA["BLI21_coassembly_000001026396_2"] <- "UNK"

```

The FERM/RESP distinction among *hgcA* sequences starts to break down with the Kiritimatiellae.
The few RESP-assigned bins are in the UBA8416 family, but not all bins in that family are RESP.
These bins are not clustered somewhere, they're distributed throughout the Kiritimatiellae *hgcA* cluster.
Think I need to abandon the distinction for these.
We'll just call them KIR. 

```{r tax_meta_assignment_by_phylogeny_continued}
# Big group of Kiritimatiellae
Kirit.seqs <- tree_subset(hgcA.tree,
                          hgcA.tree %>% MRCA(c("BLI21_coassembly_000000017500_11",
                                               "BLI21_assembly104_000000125678_2")),
                          levels_back = 0)$tip.label %>%
  strsplit(":") %>% sapply("[", 1) %>%
  grep("BLI", ., value = TRUE)
# Calling all of these Kiritimatiellae, and KIR for metabolism
taxonomic_grouping_hgcA[Kirit.seqs] <- "Kiritimatiellae"
metabolic_grouping_hgcA[Kirit.seqs] <- "KIR"
rm(Kirit.seqs)

# Four sequences, very close to Kiritimatiellae. However, three were binned as Firmicutes.
Firmi_group <- c("BLI21_coassembly_000000113360_4", "BLI21_coassembly_000000375864_1",
                 "BLI21_coassembly_000000887220_2", "BLI21_assembly105_000000019377_4") 
taxonomic_grouping_hgcA[Firmi_group] <- "Firmicutes"
metabolic_grouping_hgcA[Firmi_group] <- "FERM"
rm(Firmi_group)

# Three sequences, very close to Firmicutes group above. However, one was binned as Desulfomonilaceae.
Desulfo_group <- c("BLI21_coassembly_000000025634_9", "BLI21_coassembly_000000086306_5", "BLI21_coassembly_000000502164_1")
taxonomic_grouping_hgcA[Desulfo_group] <- "Desulfomonilaceae"
metabolic_grouping_hgcA[Desulfo_group] <- "SRB"
rm(Desulfo_group)

# Two bins close to Planctomycetes and Spirochaetes, not clearly related though.
# We'll call these generic PVC. Let's go back and call the other Spirochaetes as PVC.
# The BLI21_coassembly_000000724865_2 sequence we'll call as Planctomycetes, based on previous analysis.
PVC_group <- c("BLI21_coassembly_000000466192_1", "BLI21_coassembly_000000059411_4")
taxonomic_grouping_hgcA[PVC_group] <- "PVC"
metabolic_grouping_hgcA[PVC_group] <- "UNK"

# Slightly smaller group of Kiritimatiellae, still fairly large. Many are binned.
# The references for two are Lentisphaerae, but the bins say Kiritimatiellae.
# Many Kiritimatiellae MAGs seem to be mis-annotated as Lentisphaerae.
kirit_group_2 <- tree_subset(hgcA.tree,
                             hgcA.tree %>% MRCA(c(grep("BLI21_coassembly_000000000998_49", hgcA.tree$tip.label, value = TRUE),
                                                  grep("BLI21_coassembly_000000421460_2", hgcA.tree$tip.label, value = TRUE))),
                             levels_back = 0)$tip.label %>%
  strsplit(":") %>% sapply("[", 1) %>%
  grep("BLI", ., value = TRUE)
# Calling all of these Kiritimatiellae, and KIR for metabolism
taxonomic_grouping_hgcA[kirit_group_2] <- "Kiritimatiellae"
metabolic_grouping_hgcA[kirit_group_2] <- "KIR"
rm(kirit_group_2)

# Big group of sequences without any references, but there was a binned sequence here that was Bacteroidetes
bacteroidetes_group_2 <- tree_subset(hgcA.tree,
                                     hgcA.tree %>% MRCA(c("BLI20_coassembly_000000131407_3",
                                                          "BLI21_coassembly_000000055702_2")),
                                     levels_back = 0)$tip.label %>%
  strsplit(":") %>% sapply("[", 1) %>%
  grep("BLI", ., value = TRUE)
# Calling all of these Bacteroidetes, and RESP for metabolism
taxonomic_grouping_hgcA[bacteroidetes_group_2] <- "Bacteroidetes"
metabolic_grouping_hgcA[bacteroidetes_group_2] <- "RESP"
rm(bacteroidetes_group_2)

# Two sequences, without a close reference. One was binned as Verruco
verruco_group <- c("BLI21_assembly103_000000154963_2", "BLI21_coassembly_000000042083_4")
taxonomic_grouping_hgcA[verruco_group] <- "Verrucomicrobiae"
metabolic_grouping_hgcA[verruco_group] <- "RESP"
rm(verruco_group)

# One sequence, off on it's own.
taxonomic_grouping_hgcA["BLI21_coassembly_000000438757_1"] <- "UNK"
metabolic_grouping_hgcA["BLI21_coassembly_000000438757_1"] <- "UNK"

# hgcA sequences within a big cluster of reference sequences that are Desulfobacterota
desulfo_group_2 <- tree_subset(hgcA.tree,
                               hgcA.tree %>% MRCA(c("BLI20_assembly002_000000161823_1",
                                                    "BLI21_coassembly_000000388521_2")),
                               levels_back = 0)$tip.label %>%
  strsplit(":") %>% sapply("[", 1) %>%
  grep("BLI", ., value = TRUE)
taxonomic_grouping_hgcA[desulfo_group_2] <- "Desulfobacterota"
metabolic_grouping_hgcA[desulfo_group_2] <- "SRB"
rm(desulfo_group_2)

# hgcA sequences within a big cluster of reference sequences that are Actinobacteria
actino_group_2 <- tree_subset(hgcA.tree,
                              hgcA.tree %>% MRCA(c("BLI20_assembly001_000000004268_4",
                                                   "BLI21_coassembly_000000684452_2")),
                              levels_back = 0)$tip.label %>%
  strsplit(":") %>% sapply("[", 1) %>%
  grep("BLI", ., value = TRUE)
taxonomic_grouping_hgcA[actino_group_2] <- "Actinobacteria"
metabolic_grouping_hgcA[actino_group_2] <- "RESP"
rm(actino_group_2)
```

Let's add this to the hgcA.data table.

```{r add_tax_meta_data_to_hgcA_table}
seqID_clusterID_key <- hgcA.data$cluster_ID
names(seqID_clusterID_key) <- hgcA.data$seqID
names(taxonomic_grouping_hgcA) <- seqID_clusterID_key[names(taxonomic_grouping_hgcA)]
names(metabolic_grouping_hgcA) <- seqID_clusterID_key[names(metabolic_grouping_hgcA)]
hgcA.data <- hgcA.data %>%
  mutate(taxonomic_assignment = taxonomic_grouping_hgcA[cluster_ID],
         metabolic_assignment = metabolic_grouping_hgcA[cluster_ID])
```

I wanted to double check the assignments

```{r check_on_assignments}
hgcA_tree_check <- hgcA.tree
hgcA_tree_check$tip.label <- hgcA_tree_check$tip.label %>%
  strsplit(":") %>% sapply("[", 1)
hgcA_tree_check_vector <- paste(hgcA.data$seqID,
                                hgcA.data$taxonomic_assignment,
                                hgcA.data$metabolic_assignment,
                                sep = ":")
names(hgcA_tree_check_vector) <- hgcA.data$seqID
hgcA_tree_check$tip.label[this.study.indices] <- hgcA_tree_check_vector[hgcA_tree_check$tip.label[this.study.indices]]

# Generate tree
hgcA.tree.image <- ggtree(hgcA_tree_check) + 
  geom_tiplab(size = 2.5,
              color = color.vector) +
  geom_treescale(x = 9)
pdf("working_directory/BLI_hgcA_tree_rooted_check.pdf",
    height = 50,
    width = 8)
hgcA.tree.image +
  geom_text2(aes(subset = !isTip,
                 label = node),
             size = 1.5)
dev.off()

```

Looks good!


```{r taxonomy_by_autoClass, message = FALSE}

library(BoSSA)
library(phyloseq)
library(tidyverse)
sqlite_file <- "dataEdited/ABA/hgcA/hgcA_for_autoClass.sqlite"
jplace_file <- "dataEdited/ABA/hgcA/hgcA_for_autoClass.jplace"
pplace_object <- read_sqlite(sqlite_file,
                             jplace_file)
refpkg_path <- "references/Hg-MATE-Db.v1.ISOCELMAG_HgcA_full.refpkg"
taxonomy <- refpkg(refpkg_path,
                   type = "taxonomy")
taxonomy.table <- pplace_to_taxonomy(pplace_object,
                                     taxonomy,
                                     tax_name = TRUE,
                                     rank = c("phylum", "class", "order",
                                              "family", "genus", "species"))
taxonomy.table.df <- as.data.frame(taxonomy.table) %>%
  mutate(seqID = row.names(taxonomy.table)) %>%
  select(seqID, phylum, class, order,
         family, genus, species) %>%
  dplyr::rename(autoclass_phylum = phylum,
                autoclass_class = class,
                autoclass_order = order,
                autoclass_family = family,
                autoclass_genus = genus,
                autoclass_species = species) %>%
  mutate(autoclass_phylum = autoclass_phylum %>%
           strsplit(" <") %>% sapply("[", 1))
hgcA.data <- hgcA.data %>%
  left_join(taxonomy.table.df)
rm(sqlite_file, jplace_file, pplace_object, refpkg_path, taxonomy, taxonomy.table, taxonomy.table.df)
```


----

# *hgcB* data

```{r hgcB_data }
library(Biostrings)
library(DECIPHER)
library(tidyverse)
# Investigate alignment for hgcB hits
readAAStringSet("dataEdited/ABA/hgcA/hgcB_clean.afa",
                format = "fasta",
                nrec = -1L,
                skip = 0L,
                seek.first.rec = FALSE,
                use.names = TRUE) %>%
  BrowseSeqs(htmlFile = "dataEdited/ABA/hgcA/hgcB_raw_hits.html")
```

Six genes were sufficiently truncated that the conserved domain (CM/IECGA) was not in the gene:

- BLI20_assembly005_000000188864_1
- BLI21_coassembly_000000421460_3
- BLI21_coassembly_000000483203_1
- BLI20_assembly001_000000070404_2
- BLI21_coassembly_000001026396_1
- BLI20_coassembly_000000275727_4

For all of these, I checked the GFF file, and it was clear that the gene was at the end of a scaffold.
Therefore, I'll count all of these as *hgcB* genes.
We'll look for the possibility of other, non-annotated *hgcB* genes later.

```{r }
```

----

# Abundance data for *hgcA*

Then I processed the coverage data by normalizing to the coverage of the 16 ribosomal proteins, which are single copy core genes (SCG). See Sorek et al, 2007 for data on identification of these genes as SCG.
SCG abundance values were calculated using the SiCoGeCo workflow from [HomeBio](https://github.com/petersonben50/HomeBio).


```{r coverage_normalization, echo = 1:3, message = FALSE}
normalized.coverage.vector <- readRDS("dataEdited/scg_coverage/scg_normalization_vector.rds")
hgcA.coverage <- read.table("dataEdited/ABA/hgcA/hgcA_coverage.tsv",
                            col.names = c("scaffoldID", "coverage", "length_of_scaffold", "metagenomeID")) %>%
  mutate(NF = normalized.coverage.vector[metagenomeID],
         coverage_normalized = coverage*NF) %>%
  select(-c(coverage, NF)) %>%
  spread(key = metagenomeID,
         value = coverage_normalized)
hgcA.data <- hgcA.data %>%
  left_join(hgcA.coverage)
rm(normalized.coverage.vector,
   hgcA.coverage)

```


----

# Expression data for *hgcA*

Last, I generated expression data for each of the *hgcA* sequences. First, I searched through the pseudoalignment data for the number of metatranscriptomic reads that map to each *hgcA* sequence. This is the data that was pulled into this file.

```{r expression_normalization, echo = 1:3, message = FALSE}
NFs.expression <- readRDS("dataEdited/metatranscriptomes/normalization_vector.rds")
hgcA.expr <- read.table("dataEdited/ABA/hgcA/hgcA_MT_coverage.tsv",
                        header = TRUE) %>%
  mutate(counts_per_bp = counts/effective_length,
         million_counts_per_liter = counts_per_bp*NFs.expression[mtID]/1000000) %>%
  select(seqID, mtID, million_counts_per_liter) %>%
  spread(key = mtID,
         value = million_counts_per_liter)
hgcA.data <- hgcA.data %>%
  left_join(hgcA.expr)
rm(NFs.expression,
   hgcA.expr)

```


Since in my analyses I average the coverage of all the sequences in a cluster for each metagenome, I wanted to make sure that the sequences within a cluster have similar abundance patterns. 

----
# Save out data

```{r save_out_data}
write.csv(hgcA.data,
          "dataFinal/hgcA_data.csv",
          row.names = FALSE)

```
### References

Sorek, R., Y. Zhu, C. J. Creevey, M. P. Francino, P. Bork, and E. M. Rubin. “Genome-Wide Experimental Determination of Barriers to Horizontal Gene Transfer.” Science 318, no. 5855 (November 30, 2007): 1449–52. https://doi.org/10.1126/science.1147112.
