---
title: "hgcA_analysis"
author: "Benjamin D. Peterson"
date: "2023-04-15"
output: html_document
---

```{r setup, include = FALSE}
rm(list = ls())
#setwd("/Users/benjaminpeterson/Documents/research/BLiMMP")
#new.wd <- "/Users/benjaminpeterson/Documents/research/BLiMMP"
library(tidyverse)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(echo = TRUE)
new.wd <- gsub("\\/[[:alpha:]]+$", "", getwd())
knitr::opts_knit$set(root.dir = new.wd)
rm(new.wd)
```

```{r read_in_color_translator, include = FALSE}
cb.translator <- readRDS("references/colorblind_friendly_colors.rds")
```

# Prepare data

First we'll read in the metadata and generate a useful table to link to the abundance data.

```{r prepare_MG_MT_metadata}
MG_metadata <- read.csv("metadata/metagenome_metadata.csv") %>%
  dplyr::rename(omicID = metagenomeID) %>%
  mutate(seqType = "MG") %>%
  select(omicID, sampleID, startDate, depth, seqType)
MT_metadata <- read.csv("metadata/metatranscriptome_metadata.csv") %>%
  dplyr::rename(omicID = metatranscriptomeID) %>%
  mutate(seqType = "MT") %>%
  select(omicID, sampleID, startDate, depth, seqType)
omic_metadata <- rbind(MG_metadata,
                       MT_metadata) %>%
  mutate(date_depth = paste(startDate, ":", depth, "m",
                            sep = ""))
rm(MG_metadata, MT_metadata)
```

Then we'll read in and prepare the *hgcA* data.
Essentially, here we'll be aggregating the MG and MT coverage by metabolic group and overall coverage for each metagenome/metatranscriptome.

```{r prepare_hgcA_data, message = FALSE}
hgcA_abundance <- read.csv("dataFinal/hgcA_data.csv")

# There were no MT hits to hgcA in BLI21_MT_003, which excluded it from the table.
hgcA_abundance$BLI21_MT_003 <- 0

# Group the coverage by the metabolic code assigned to each sequence.
hgcA_abundance_grouped <- hgcA_abundance %>%
  filter(verified_hgcA,
         clstr_rep == 1) %>%
  select(metabolic_assignment, all_of(grep("BLI2", names(hgcA_abundance)))) %>%
  gather(key = omicID, value = omic_coverage, -1) %>%
  group_by(omicID, metabolic_assignment) %>%
  summarise(omic_coverage = sum(omic_coverage)) %>%
  ungroup() %>%
  left_join(omic_metadata) %>%
  mutate(omic_type = omicID %>% strsplit("_") %>% sapply("[", 2)) %>%
  group_by(omic_type, date_depth, metabolic_assignment) %>%
  summarise(omic_coverage = median(omic_coverage)) %>%
  ungroup()

# Calculate the total hgcA coverage in the MGs and MTs.
hgcA_abundance_total <- hgcA_abundance_grouped %>%
  group_by(omic_type, date_depth) %>%
  summarise(omic_coverage = sum(omic_coverage)) %>%
  spread(key = omic_type,
         value = omic_coverage)

# Clean up abundance data for each guild in each omic type
hgcA_abundance_guild <- hgcA_abundance_grouped %>%
  mutate(guild_abund = paste(omic_type, "_", metabolic_assignment,
                             sep = "")) %>%
  select(date_depth, guild_abund, omic_coverage) %>%
  spread(key = guild_abund,
         value = omic_coverage)

# Generate final data frame and clean up
hgcA_abundance_final <- full_join(hgcA_abundance_total,
                                  hgcA_abundance_guild)
rm(hgcA_abundance_grouped, hgcA_abundance_guild,
   hgcA_abundance_total, hgcA_abundance)
```

We'll also want to pull in the sulfide data and the K~met~ values from the incubations.
In particular, we'll want to calculate an "SRB-dependent" Kmet value, which is the Kmet value of the ambient incubations minus the Kmet value of the molybdate-amended incubations.

```{r prepare_kmet_incubation_sulfide_data, message = FALSE}
# Read in incubation rate data
Hg_Kmet_data <- read.csv("dataFinal/incubation_Hg_rate_data.csv") %>%
  dplyr::rename(date = startDate) %>%
  filter(treatment != "filtered-unamended",
         year(date) %in% c(2020, 2021)) %>%
  mutate(treatment = treatment %>%
           gsub("unfiltered-", "", .) %>%
           gsub("unamended", "ambient", .)) %>%
  select(sampleID, date, depth, treatment, Kmet_int_t2) 

# Read in sulfide data
sulfide_data <- read.csv("dataFinal/water_chem_data.csv") %>%
  group_by(date, depth) %>%
  summarize(sulfide_uM = mean(sulfide_uM, na.rm = TRUE)) %>%
  ungroup() %>%
  mutate(date_depth = paste(date, ":", depth, "m",
                            sep = ""))

# Calculate SRB-dependent Kmet
Hg_Kmet_modified <- Hg_Kmet_data %>%
  group_by(sampleID, date, depth, treatment) %>%
  summarise(Kmet_mean = median(Kmet_int_t2),
            Kmet_sd = sd(Kmet_int_t2),
            Kmet_count = n(),
            Kmet_se = Kmet_sd / Kmet_count) %>%
  ungroup() %>%
  gather(key = constituent,
         value = value,
         -c(1:4)) %>%
  mutate(constituent = paste(treatment, "_", constituent,
                             sep = "")) %>%
  select(sampleID, date, depth, constituent, value) %>%
  spread(key = constituent,
         value = value) %>%
  mutate(SRB_Kmet_mean = ambient_Kmet_mean - molybdate_Kmet_mean,
         SRB_Kmet_se = ambient_Kmet_se + molybdate_Kmet_se) %>%
  dplyr::rename("non-SRB_Kmet_mean" = molybdate_Kmet_mean,
         "non-SRB_Kmet_se" = molybdate_Kmet_se,
         total_Kmet_mean = ambient_Kmet_mean,
         total_Kmet_se = ambient_Kmet_se) 


# Join sulfide and incubation rate potential data
Hg_Kmet_modified <- Hg_Kmet_modified %>%
  left_join(sulfide_data) %>%
  mutate(date_depth = paste(date, ":", depth, "m",
                            sep = "")) %>%
  select(date, depth, date_depth, sulfide_uM, total_Kmet_mean, total_Kmet_se, SRB_Kmet_mean, SRB_Kmet_se, "non-SRB_Kmet_mean", "non-SRB_Kmet_se") %>%
  gather(key = measurement,
         value = Kmet_value,
         -c(1:4)) %>%
  mutate(guild = measurement %>%
           strsplit("_") %>% sapply("[", 1),
         measurement = paste(measurement %>%
                               strsplit("_") %>% sapply("[", 2),
                             measurement %>%
                               strsplit("_") %>% sapply("[", 3),
                             sep = "_")) %>%
  spread(key = measurement,
         value = Kmet_value) %>%
  filter(!(date == "2020-10-10" & depth == 15.7))


```

----

# Total *hgcA* values


First, let's just plot the methylation rate constant against the sulfide concentrations, metagenomic abundance of *hgcA*, and expression of *hgcA*.

```{r Kmet_vs_hgcA, echo = FALSE, message = FALSE, fig.width = 7.5, fig.height = 7.5 }
year_vector <- c(21, 24)
names(year_vector) <- c("2020", "2021")

Kmet_vs_sulfide <- function() {
  plot_data <- Hg_Kmet_modified %>%
    filter(guild == "total",
           !is.na(Kmet_mean))
  plot(x = plot_data$sulfide_uM,
       y = plot_data$Kmet_mean,
       xlim = c(0, 150),
       ylim = c(0, 0.2),
       pch = year_vector[as.character(year(plot_data$date))],
       col = "gray25",
       bg = "gray85",
       lwd = 1.5,
       cex = 2,
       ylab = expression('K'['met']*' (day'^-1*')'),
       xlab = "Sulfide (µM)")
}
Kmet_vs_hgcA_MG <- function() {
  plot_data <- left_join(hgcA_abundance_final,
                         Hg_Kmet_modified %>%
                           filter(guild == "total")) %>%
    filter(!is.na(Kmet_mean))
  plot(x = plot_data$MG,
       y = plot_data$Kmet_mean,
       xlim = c(0, 17),
       ylim = c(0, 0.2),
       pch = year_vector[as.character(year(plot_data$date))],
       col = "gray25",
       bg = "gray85",
       lwd = 1.5,
       cex = 2,
       ylab = expression('K'['met']*' (day'^-1*')'),
       xlab = "hgcA gene abundance (%)")
}
Kmet_vs_hgcA_MT <- function() {
  plot_data <- left_join(hgcA_abundance_final,
                         Hg_Kmet_modified %>%
                           filter(guild == "total")) %>%
    filter(!is.na(Kmet_mean),
           !is.na(MT))
  plot(x = plot_data$MT,
       y = plot_data$Kmet_mean,
       xlim = c(0, 12),
       ylim = c(0, 0.2),
       pch = year_vector[as.character(year(plot_data$date))],
       col = "gray25",
       bg = "gray85",
       lwd = 1.5,
       cex = 2,
       ylab = expression('K'['met']*' (day'^-1*')'),
       xlab = "hgcA gene transcription (million transcripts per L)")
  legend(x = 8,
         y = 0.05,
         legend = names(year_vector),
         pch = year_vector,
         col = "gray25",
         pt.bg = "gray85",
         lwd = 1.5,
         cex = 1)
}

hgcA_MT_vs_hgcA_MG <- function() {
  plot_data <- left_join(hgcA_abundance_final,
                         Hg_Kmet_modified %>%
                           filter(guild == "total")) %>%
    filter(!is.na(MG),
           !is.na(MT))
  plot(x = log(plot_data$MG, 10),
       y = log(plot_data$MT, 10),
       # xlim = c(0, 17),
       # ylim = c(0, 12),
       pch = year_vector[as.character(year(plot_data$date))],
       col = "gray25",
       bg = "gray85",
       lwd = 1.5,
       cex = 2,
       xlab = "Log10 hgcA gene abundance (%)",
       ylab = "Log10 hgcA gene transcription (million transcripts per L)")
  legend(x = 8,
         y = 0.05,
         legend = names(year_vector),
         pch = year_vector,
         col = "gray25",
         pt.bg = "gray85",
         lwd = 1.5,
         cex = 1)
}

#### Set up plots ####
par(mfrow = c(2, 2),
    mar = c(3, 3, 1, 1),
    tck = -0.008,
    mgp = c(1.5, 0.2, 0))
Kmet_vs_sulfide()
Kmet_vs_hgcA_MG()
Kmet_vs_hgcA_MT()
hgcA_MT_vs_hgcA_MG()

#### Generate table for Kmet and sulfide ####
knitr::kable(left_join(hgcA_abundance_final,
                         Hg_Kmet_modified %>%
                           filter(guild == "total")) %>%
               select(date_depth, sulfide_uM, Kmet_mean, MG, MT) %>%
               mutate(Kmet_mean = round(Kmet_mean, 3),
                      sulfide_uM = round(sulfide_uM, 1),
                      MG = round(MG, 1),
                      MT = round(MT, 1)) %>%
               arrange(desc(Kmet_mean)),
             col.names = c("Date and depth", "Sulfide (µM)", 'Kmet (day^-1)', "hgcA gene abundance (%)", "hgcA gene transcription (million counts per liter)"))
```

First, in the Kmet vs. sulfide pot, we see that methylation in this system follows the Goldilocks curve, for the most part.
The data is a bit noisy to say exactly when that happens, but peak methylation occurs around 70-100 µM sulfide, then starts to tail off.
Could be some interannual differences at play here too, as we see the peak occur at ~70 µM sulfide in 2020 and 122 µM sulfide in 2021.
This could be due to factors such as carbon composition influencing the ligand chemistry and therefore bioavailability.
Temperature could be a factor, although this seems similar between years.
Alternatively, it could be linked to the changes in *hgcA* abundance between years, since we don't understand what controls abundance.

We see a similar trend with the *hgcA* gene abundance in relation to K~met~.
I suspect that this is because *hgcA* gene abundance tracks closely with sulfide.
However, with the metatranscriptomes, we see a more linear relationship between total *hgcA* expression and K~met~.
This indicates several possibilities, in comparing the MGs to the MTs:
1. The metatranscriptomes are more effective at predicting K~met~ simply because there is an internal standard and we can calculate an absolute number of transcripts.
2. The *hgcA* transcription is related to the bioavailability of Hg(II). I find this unlikely but possible.

---

# Functional guilds of *hgcA*

The next objective of this project was to look at the various functional guilds of mercury methylators and the drivers of MeHg production in these incubations.
To do that, I'm plotting here the influence of molybdate on K~met~ as well as the functional guild composition of *hgcA* genes and gene transcription. 
All these will be plotted as a function of sulfide for comparison.

```{r functional_guilds_of_hgcA, echo = FALSE, message = FALSE, fig.width = 9, fig.height = 3  }
color_vector <- c(cb.translator[c("black", "blue", "yellow", "bluishgreen")], "gray")
names(color_vector) <- c("KIR", "SRB", "RESP", "FERM", "UNK")
name_vector <- c(expression('K'['met']*' (day'^-1*')'),
                'SRB-independent',
                'SRB-dependent')
# Kmet
plot_Kmet_vs_sulfide_guilds <- function() {
  total_Kmet_data <- Hg_Kmet_modified %>%
    filter(guild == "total")
  plot(x = total_Kmet_data$sulfide_uM,
       y = total_Kmet_data$Kmet_mean,
       xlim = c(0, 150),
       ylim = c(0, 0.2),
       pch = year_vector[as.character(year(total_Kmet_data$date))],
       col = "gray25",
       bg = "gray85",
       lwd = 1.5,
       cex = 2,
       xlab = "Sulfide (µM)",
       ylab = expression('K'['met']*' (day'^-1*')'))
  nonSRB_Kmet_data <- Hg_Kmet_modified %>%
    filter(guild == "non-SRB")
  points(x = nonSRB_Kmet_data$sulfide_uM,
         y = nonSRB_Kmet_data$Kmet_mean,
         pch = year_vector[as.character(year(nonSRB_Kmet_data$date))],
         cex = 1.2,
         col = "gray50",
         bg = "gray",
         lwd = 2)
  SRB_Kmet_data <- Hg_Kmet_modified %>%
    filter(guild == "SRB")
  points(x = SRB_Kmet_data$sulfide_uM,
         y = SRB_Kmet_data$Kmet_mean,
         pch = year_vector[as.character(year(SRB_Kmet_data$date))],
         cex = 1.2,
         col = "gray50",
         bg = color_vector["SRB"],
         lwd = 2)
  legend(x = -2,
         y = 0.18,
         legend = name_vector,
         pch = c(21, 16, 16),
         pt.cex = c(2, 1.2, 1.2),
         pt.lwd = 2,
         col = c("gray25", "gray", color_vector["SRB"]),
         pt.bg = "gray85",
         bty = "n")
}

omics_points_function <- function(omicType = "MG", guild = "KIR") {
  plot_data <- left_join(hgcA_abundance_final,
                         sulfide_data) %>%
    as.data.frame()
  column_name <- paste(omicType, guild, sep = "_")
  points(x = plot_data$sulfide_uM,
         y = plot_data[, column_name],
         pch = year_vector[as.character(year(plot_data$date))],
         col = "gray25",
         bg = color_vector[guild],
         lwd = 1.5,
         cex = 1)}
# Metagenomes
hgcA_MG_plot <- function() {
  plot_data <- left_join(hgcA_abundance_final,
                         sulfide_data)
  plot(x = plot_data$sulfide_uM,
       y = plot_data$MG,
       xlim = c(0, 150),
       ylim = c(0, 20),
       pch = year_vector[as.character(year(plot_data$date))],
       col = "gray25",
       bg = "gray85",
       lwd = 2,
       cex = 2,
       xlab = "Sulfide (µM)",
       ylab = "hgcA gene coverage")
  sapply(names(color_vector),
         function(guild_to_plot) {
           omics_points_function("MG", guild_to_plot)
         })
  }

# Metatranscriptomes
hgcA_MT_plot <- function() {
  plot_data <- left_join(hgcA_abundance_final,
                         sulfide_data)
  plot(x = plot_data$sulfide_uM,
       y = plot_data$MT,
       xlim = c(0, 150),
       ylim = c(0, 12),
       pch = year_vector[as.character(year(plot_data$date))],
       col = "gray25",
       bg = "gray85",
       lwd = 2,
       cex = 2,
       xlab = "Sulfide (µM)",
       ylab = "hgcA expression (million copies/L)")
  sapply(names(color_vector),
         function(guild_to_plot) {
           omics_points_function("MT", guild_to_plot)
         })
  legend(x = 100,
         y = 10,
         legend = names(color_vector),
         pch = c(16),
         col = color_vector,
         pt.cex = 2,
         bty = "n")
}

#### Lay out plots ####
par(mfrow = c(1, 3),
    mar = c(3, 3, 1, 1),
    tck = -0.008,
    mgp = c(1.5, 0.2, 0))
plot_Kmet_vs_sulfide_guilds()
hgcA_MG_plot()
hgcA_MT_plot()

```

Does the abundance and/or transcription of SRB-associated *hgcA* correlate to the influence of molybdate on methylation rates?

```{r SRB_dependence, message = FALSE, echo = FALSE }
hgcA_percent_SRB <- left_join(Hg_Kmet_modified %>%
                                select(date_depth, sulfide_uM, guild, Kmet_mean) %>%
                                spread(key = guild,
                                       value = Kmet_mean) %>%
                                mutate(SRB_per_Kmet = round(SRB / total * 100, 1),
                                       sulfide_uM = round(sulfide_uM, 1)) %>%
                                select(date_depth, sulfide_uM, SRB_per_Kmet),
                              hgcA_abundance_final %>%
                                mutate(SRB_per_MT = round(MT_SRB / MT * 100, 1),
                                       SRB_per_MG = round(MG_SRB / MG * 100, 1)) %>%
                                select(date_depth, SRB_per_MG, SRB_per_MT)) %>%
  filter(date_depth != "2021-09-10:10.8m") %>%
  arrange(SRB_per_Kmet) %>%
  as.data.frame()
par(mfrow = c(1, 2),
    mar = c(3, 3, 1, 1),
    tck = -0.008,
    mgp = c(1.5, 0.2, 0))
plot(x = hgcA_percent_SRB$SRB_per_MG,
     y = hgcA_percent_SRB$SRB_per_Kmet,
     xlab = "Percent SRB-associated hgcA in MG",
     ylab = "Percent SRB-associated Kmet",
     pch = 16,
     xlim = c(0, 100),
     ylim = c(0, 100))
plot(x = hgcA_percent_SRB$SRB_per_MT,
     y = hgcA_percent_SRB$SRB_per_Kmet,
     xlab = "Percent SRB-associated hgcA in MT",
     ylab = "Percent SRB-associated Kmet",
     pch = 16,
     xlim = c(0, 100),
     ylim = c(0, 100))
hgcA_percent_SRB
```

Not really, but the relative abundance of SRB-associated *hgcA* genes in the metatranscriptomes is closer to the relative inhibition of Kmet by molybdate.


# Drivers of increased SRB *hgcA* transcription

Are the SRBs in general more transcriptionally active, or is it specific to the *hgcA* gene?
Here, we'll hone in on the bins that have *hgcA* rather than the total of the assembly-derived *hgcA* sequences and use both *gyrB* and total mRNA reads to compare to *hgcA*.
First read in the data.

```{r read_in_and_prepare_bin_MT_data, warning=FALSE, message=FALSE}
bin_metadata <- readRDS(file = 'dataEdited/bin_based_analyses/bin_data_aggregate_3.rds')
```

Next, plot the RNA levels from each metabolic group of *hgcA*, *gyrB*, and total RNA. We'll do this against sulfide, as we did for the assembly-derived *hgcA* genes.

```{r plot_RNA_data, warning = FALSE, message = FALSE, echo = FALSE, fig.width = 8, fig.height = 3}
sulfide_data_for_omics <- omic_metadata %>%
  select(omicID, date_depth) %>%
  filter(grepl("_MT_", omicID)) %>%
  rename(mt_id = omicID) %>%
  left_join(sulfide_data %>%
              select(date_depth, sulfide_uM))

constituent <- "hgcA"
# Metatranscriptomes
plot_bin_MT_by_sulfide <- function(constituent,
                                   y_scale_to_use,
                                   y_label_to_use = "testing") {
  # Add KIR notation to meta_code
  plot_data <- bin_metadata
  plot_data$meta_code[bin_metadata$class == "Kiritimatiellae"] <- "KIR"
  plot_data$meta_code <- gsub(pattern = "OB_",
                              replacement = "",
                              plot_data$meta_code)
  plot_data <- plot_data %>%
    select(meta_code, hms_id, all_of(grep("_MT_",
                                          names(plot_data),
                                          value = TRUE))) %>%
    gather(key = gene_mt_id,
           value = gene_transcription,
           all_of(grep("_MT_",
                       names(plot_data),
                       value = TRUE))) %>%
    mutate(gene = gene_mt_id %>% strsplit("-") %>% sapply("[", 1),
           mt_id = gene_mt_id %>% strsplit("-") %>% sapply("[", 2)) %>%
    select(-gene_mt_id) %>%
    filter(!is.na(gene_transcription),
           gene == constituent) %>%
    group_by(meta_code, hms_id, mt_id) %>%
    summarise(gene_transcription = mean(gene_transcription))
  
  # Set up proper constituent
  plot_data <- left_join(plot_data,
                         sulfide_data_for_omics)
  plot_data <- plot_data %>%
    group_by(meta_code, date_depth, mt_id, sulfide_uM) %>%
    summarise(gene_transcription = sum(gene_transcription,
                                       na.rm = TRUE)) %>%
    ungroup() %>%
    group_by(meta_code, date_depth, sulfide_uM) %>%
    summarise(gene_transcription = mean(gene_transcription,
                                        na.rm = TRUE))

  # Generate plots
  plot(x = plot_data$sulfide_uM,
       y = plot_data$gene_transcription,
       xlim = c(0, 150),
       lwd = 2,
       cex = 2,
       pch = 21,
       col = "gray20",
       bg = color_vector[plot_data$meta_code],
       xlab = "Sulfide (µM)",
       ylab = y_label_to_use)
}
par(mfrow = c(1, 3),
    mar = c(3, 3, 1, 1),
    tck = -0.008,
    mgp = c(1.5, 0.2, 0))

plot_bin_MT_by_sulfide(constituent = "hgcA",
                       y_label_to_use = "hgcA expression (million copies/L)")
plot_bin_MT_by_sulfide(constituent = "gyrB",
                       y_label_to_use = "gyrB expression (million copies/L)")
plot_bin_MT_by_sulfide(constituent = "total",
                       y_label_to_use = "Total RNA expression (million copies/L)")
legend(x = 0,
       y = 70000,
       legend = names(color_vector),
       pch = c(16),
       col = color_vector,
       pt.cex = 2,
       bty = "n")
```

Overall, these match up relatively well across the different metrics.
Notably, the KIR hgcA+ genomes account for a higher percent of the *hgcA* expression relative to SRB-associated genomes than observed for the assembly-based analyses.
This is likely due to the increased recovery of the KIR genomes, since they were far higher in abundance in the metagenomes.

To isolate this from the effect of increased bin recovery, I'll compare the *hgcA* expression to *gyrB* within the same genome.

```{r plot_RNA_data_hgcA_gyrB, warning = FALSE, message = FALSE, echo = FALSE, fig.width = 9, fig.height = 4}
plot_bin_MT_genes <- function(constituent_1,
                              constituent_2,
                              label_1,
                              label_2,
                              xlim_to_use = NULL,
                              ylim_to_use = NULL) {
  # Add KIR notation to meta_code
  plot_data <- bin_metadata
  plot_data$meta_code[bin_metadata$class == "Kiritimatiellae"] <- "KIR"
  plot_data$meta_code <- gsub(pattern = "OB_",
                              replacement = "",
                              plot_data$meta_code)
  plot_data_1 <- plot_data %>%
    select(meta_code, hms_id, all_of(grep("_MT_",
                                          names(plot_data),
                                          value = TRUE))) %>%
    gather(key = gene_mt_id,
           value = gene_transcription,
           all_of(grep("_MT_",
                       names(plot_data),
                       value = TRUE))) %>%
    mutate(gene = gene_mt_id %>% strsplit("-") %>% sapply("[", 1),
           mt_id = gene_mt_id %>% strsplit("-") %>% sapply("[", 2)) %>%
    select(-gene_mt_id) %>%
    filter(!is.na(gene_transcription)) %>%
    group_by(meta_code, gene, hms_id, mt_id) %>%
    summarise(gene_transcription = mean(gene_transcription)) %>%
    spread(key = gene,
           value = gene_transcription) 
  if ("gyrB" %in% c(constituent_1, constituent_2)) {
    plot_data_1 <- plot_data_1 %>%
      filter(gyrB > 0)
  }
  if ("hgcA" %in% c(constituent_1, constituent_2)) {
    plot_data_1$hgcA[plot_data_1$hgcA == 0] <- 0.004
  }
  plot_data_1 <- plot_data_1 %>% as.data.frame()
  paste(length(log(plot_data_1[, constituent_1], 10)),
        length(log(plot_data_1[, constituent_2], 10)),
        length(color_vector[plot_data_1$meta_code]))
  
    plot(x = log(plot_data_1[, constituent_1], 10),
         y = log(plot_data_1[, constituent_2], 10),
         xlim = xlim_to_use,
         ylim = ylim_to_use,
         xlab = label_1,
         ylab = label_2,
         col = color_vector[plot_data_1$meta_code],
         pch = 18,
         cex = 0.8)
    }
  
par(mfrow = c(1, 3),
    mar = c(4.5, 4.5, 1, 1),
    tck = -0.008,
    mgp = c(2, 0.2, 0))

plot_bin_MT_genes(constituent_1 = "gyrB",
                  constituent_2 = "hgcA",
                  label_1 = "gyrB transcription\n(log million copies/L)",
                  label_2 = "hgcA transcription\n(log million copies/L)",
                  xlim_to_use = c(log(0.004, 10), 1),
                  ylim_to_use = c(log(0.004, 10), 1))
abline(a = 0, b = 1)
plot_bin_MT_genes(constituent_1 = "total",
                  constituent_2 = "hgcA",
                  label_1 = "Total RNA\n(log million copies/L)",
                  label_2 = "hgcA transcription\n(log million copies/L)")
plot_bin_MT_genes(constituent_1 = "total",
                  constituent_2 = "gyrB",
                  label_1 = "Total RNA\n(log million copies/L)",
                  label_2 = "gyrB transcription\n(log million copies/L)")
```

Let's save out the two *hgcA* plots to use in the manuscript:

```{r save_out_transcription_plots}

```

----


# 

