---
title: "hgcA gene neighborhood analyses"
author: "Benjamin D. Peterson"
date: "2024-01-05"
output: html_document
---

```{r setup, include=FALSE}
#rm(list = ls())
#setwd("/Users/benjaminpeterson/Documents/research/BLiMMP")
#new.wd <- "/Users/benjaminpeterson/Documents/research/BLiMMP"
library(tidyverse)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(echo = TRUE)
new.wd <- gsub("\\/[[:alpha:]]+$", "", getwd())
knitr::opts_knit$set(root.dir = new.wd)
rm(new.wd)
```


This document contains my analyses pertaining to the neighborhood of *hgcA* genes in the metagenomes from the Mendota incubations project. I will use the gene neighborhood information for three purposes:

1. Confirm taxonomic affiliation of *hgcA*-carrying organisms
2. Identify possible transcriptional regulators to relate to *hgcA* transcription.
3. Look for phylogenetically conserved trends in gene neighborhood data.

# Read in data

First we'll read in the data.

```{r read_in_hgcA_GN_backbone }
hgcA_gff <- read.table("dataEdited/ABA/hgcA/GN/hgcA_geneNeighborhood_raw.gff",
                       col.names = c("scaffoldID", "delete1", "delete2", "gene_start",
                                     "gene_end", "delete3", "gene_direction", "delete4",
                                     "attributes")) %>%
  mutate(seqID = paste(scaffoldID,
                       attributes %>% strsplit(";") %>% sapply("[", 1) %>%
                         strsplit("_") %>% sapply("[", 2),
                       sep = "_")) %>%
  select(seqID, scaffoldID, gene_start, gene_end, gene_direction)
```

This `hgcA_gff` file will serve as the foundation for this analysis, as I'll add gene information to this dataframe.

Then we'll read in the gene information. We'll classify our gene sequencing using four sources of data:
1. *hgcA* identification methods.
2. *hgcB* identification methods.
3. KOFAMscan results.
4. NCBI BLAST results.

# *hgcA*

Let's start with the *hgcA* data.

```{r read_in_hgcA_data }
hgcA_data <- read.csv("dataFinal/hgcA_data.csv") %>%
  filter(verified_hgcA) %>%
  select(seqID, cluster_ID, clstr_rep, scaffoldID,
         taxonomic_assignment, metabolic_assignment)
```

First, some housekeeping: I'm only going to deal with the verified *hgcA* sequences here, so let's filter that out.
I'll also add some basic information about the HMS cluster, taxonomy, and putative metabolism of each hgcA+ bin and will mark which genes are *hgcA*.

```{r only_keep_GN_of_verified_hgcA, message = FALSE }
hgcA_gff <- hgcA_gff %>%
  filter(scaffoldID %in% hgcA_data$scaffoldID) %>%
  left_join(hgcA_data %>% select(cluster_ID, scaffoldID, taxonomic_assignment, metabolic_assignment))
# Identify hgcA genes
hgcA_gff$gene_info <- NA
hgcA_gff$gene_info[which(hgcA_gff$seqID %in% hgcA_data$seqID)] <- "hgcA"
```

# *hgcB*

Next add in the *hgcB* data.

```{r read_in_hgcB_data }
library(Biostrings)
hgcB_data <- readAAStringSet("dataEdited/ABA/hgcA/GN/hgcB_clean.afa")
hgcA_gff$gene_info[which(hgcA_gff$seqID %in% names(hgcB_data))] <- "hgcB"
```

# KOFAMscan and NCBI BLAST results.

Next we'll start assigning annotations to the other genes in the neighborhood.
I'll add together these results and curate them manually.
First prep the kofamscan data.

```{r read_in_kofamscan_data, message = FALSE }
kofamscan_data <- read.table("dataEdited/ABA/hgcA/GN/hgcA_GN_kofamscan_qualityHits_clean.tsv",
                             sep = '\t',
                             header = TRUE) %>%
  dplyr::rename(seqID = gene.name,
                gene_info = KO.definition) %>%
  select(seqID, E.value, gene_info)
kofamscan_data$hgc_genes <- NA
kofamscan_data$hgc_genes[which(kofamscan_data$seqID %in% names(hgcB_data))] <- "hgcB"
kofamscan_data$hgc_genes[which(kofamscan_data$seqID %in% hgcA_data$seqID)] <- "hgcA"
kofamscan_data <- kofamscan_data %>% select(seqID, hgc_genes, E.value, gene_info)
```

Then the BLAST data.
I saved the top 5 hits when I performed the blast, we'll just concatenate the unique strings by group to have one entry per hit.

```{r read_in_blast_data, message = FALSE }
blast_metadata <- read.table("dataEdited/ABA/hgcA/GN/hgcA_GN_blast_refseq_data.tsv",
                             sep = '\t',
                             col.names = c("accession_id", "gene_info", "taxid"),
                             quote = "") %>%
  mutate(gene_type = gene_info %>% strsplit(" \\[") %>% sapply("[", 1),
         taxa = gene_info %>% strsplit("\\[") %>% sapply("[", 2) %>% gsub("\\]", "", .)) %>%
  select(accession_id, gene_type, taxa)
blast_data <- read.table("dataEdited/ABA/hgcA/GN/hgcA_GN_blast_data.tsv",
                         col.names = c("seqID", "score", "ID")) %>%
  mutate(accession_id = ID %>% strsplit("\\|ref\\|") %>% sapply("[", 2) %>% gsub("\\|", "", .)) %>%
  select(seqID, accession_id, score) %>%
  left_join(blast_metadata) %>%
  filter(!is.na(gene_type)) %>%
  mutate(taxa_score = paste(taxa, score, sep = "-")) %>%
  group_by(seqID) %>%
  summarise(BLAST_gene_type = paste0(unique(gene_type), collapse = ";"),
            BLAST_taxa_score = paste0(unique(taxa_score), collapse = ";"))
rm(blast_metadata)
```

Now join these together and add some info about the corresponding *hgcA* sequences.
We'll only work on the gene neighborhoods of verified *hgcA* sequences.

```{r join_blast_kofamscan_data, message = FALSE}
annotation_data <- full_join(kofamscan_data,
                             blast_data) %>%
  mutate(scaffoldID = seqID %>% strsplit("_[0-9]*$") %>% sapply("[", 1)) %>%
  left_join(hgcA_data %>% select(cluster_ID, scaffoldID, taxonomic_assignment, metabolic_assignment)) %>%
  filter(!is.na(cluster_ID))
rm(kofamscan_data, blast_data)
```

Okay, now we can start to annotate.

# Transcriptional regulators

Quick way to identify *asrR* genes:

```{r transc_reg_arsR, message = FALSE}
arsR_check_data <- annotation_data %>%
  filter(grepl("ArsR", BLAST_gene_type, ignore.case = TRUE))
hgcA_gff$gene_info[which(hgcA_gff$seqID %in% unique(arsR_check_data$seqID))] <- "transc_reg_arsR"
```

In searching for "transcription" in the BLAST_gene_type, I identified a number of other potential transcriptional regulators. We have:
- GntR (just one)
- helix-turn-helix transcriptional regulator
- MarR (just one)
- sigma 54-interacting


```{r other_transcriptional_regulators, message = FALSE}
GntR <- annotation_data %>%
  filter(grepl("GntR", BLAST_gene_type, ignore.case = TRUE)) %>%
  select(seqID) %>% unlist(use.names = FALSE)
hgcA_gff$gene_info[which(hgcA_gff$seqID %in% GntR)] <- "transc_reg_gntR"

MarR <- annotation_data %>%
  filter(grepl("MarR", BLAST_gene_type, ignore.case = TRUE)) %>%
  select(seqID) %>% unlist(use.names = FALSE)
hgcA_gff$gene_info[which(hgcA_gff$seqID %in% MarR)] <- "transc_reg_marR"

helix_turn_helix <- annotation_data %>%
  filter(grepl("helix-turn-helix", BLAST_gene_type, ignore.case = TRUE)) %>%
  select(seqID) %>% unlist(use.names = FALSE)
hgcA_gff$gene_info[which(hgcA_gff$seqID %in% helix_turn_helix)] <- "transc_reg_HTH"

transc_reg_sig54 <- annotation_data %>%
  filter(grepl("sigma[ ,-]54", BLAST_gene_type, ignore.case = TRUE)) %>%
  select(seqID) %>% unlist(use.names = FALSE)
hgcA_gff$gene_info[which(hgcA_gff$seqID %in% transc_reg_sig54)] <- "transc_reg_sig54"

rm(GntR, MarR, helix_turn_helix, transc_reg_sig54)
```

The *gntR* gene preceeds *hgcA* in one of the Actinobacteria sequences, but is well upstream.
The MarR gene is found preceeding one of the Desulfomonilaceae *hgcA* genes.
The other two are preceeded by *arsR*.

The helix-turn-helix is in only one *hgcA* cluster, far from the *hgcA* and not relevant.

The sigma-54 interaction factor in Desulfobacter is directly downstream of the *hgcB* or one ORF down, in the reverse direction.
When it's found in Kiritimatiellae, it's still downstream of *hgcB*, but much further, enough that it's not on the scaffold sometimes, and it is in the same direction.

# Arsenic-processing genes

```{r ACR3_arsenite_efflux, message = FALSE}
arsenite_efflux <- unique(annotation_data %>%
                            filter(grepl("ACR3", BLAST_gene_type, ignore.case = TRUE)) %>%
                            select(seqID) %>% unlist(use.names = FALSE),
                          annotation_data %>%
                            filter(grepl("arsenite transporter", gene_info, ignore.case = TRUE)) %>%
                            select(seqID) %>% unlist(use.names = FALSE))
hgcA_gff$gene_info[which(hgcA_gff$seqID %in% arsenite_efflux)] <- "ACR3_arsenite_efflux"
```


```{r ArsC_arsenate_reductase, message = FALSE}
arsenate_reductase <- unique(annotation_data %>%
                            filter(grepl("arsenate reductase ArsC", BLAST_gene_type, ignore.case = TRUE)) %>%
                            select(seqID) %>% unlist(use.names = FALSE))
hgcA_gff$gene_info[which(hgcA_gff$seqID %in% arsenate_reductase)] <- "ArsC_arsenate_reductase"
```


# Heavy metal resistance and other efflux proteins 

There's also a number of heavy metal efflux proteins.
A subset of these are found in pairs, with a "CusA/CzcA family heavy metal efflux RND transporter" and a "efflux RND transporter periplasmic adaptor subunit".
The CusA transporter is part of the RNA antimicrobial resistance complex known to be specific to heavy metals. 
More information here: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4622206/
We'll call the CusA-annotations "metal_efflux_RND_CusA" and the other member of the pair "metal_efflux_RND_adaptor".
There are also several TolC annotations mixed in with "efflux RND transporter permease" or "outer membrane protein, heavy metal efflux system" annotations.
We'll call these metal_efflux_RND_TolC.

```{r metal_efflux_RND_CusA, message = FALSE}
metal_efflux_RND_CusA <- unique(annotation_data %>%
                                         filter(grepl("CusA/CzcA", BLAST_gene_type, ignore.case = TRUE)) %>%
                                         select(seqID) %>% unlist(use.names = FALSE))
hgcA_gff$gene_info[which(hgcA_gff$seqID %in% metal_efflux_RND_CusA)] <- "metal_efflux_RND_CusA"
rm(metal_efflux_RND_CusA)

metal_efflux_RND_adaptor <- unique(annotation_data %>%
                                     filter(grepl("efflux RND transporter periplasmic adaptor subunit", BLAST_gene_type, ignore.case = TRUE)) %>%
                                         select(seqID) %>% unlist(use.names = FALSE))
hgcA_gff$gene_info[which(hgcA_gff$seqID %in% metal_efflux_RND_adaptor)] <- "metal_efflux_RND_adaptor"
rm(metal_efflux_RND_adaptor)

metal_efflux_RND_TolC <- unique(annotation_data %>%
                                     filter(grepl("TolC", BLAST_gene_type, ignore.case = TRUE)) %>%
                                         select(seqID) %>% unlist(use.names = FALSE))
hgcA_gff$gene_info[which(hgcA_gff$seqID %in% metal_efflux_RND_TolC)] <- "metal_efflux_RND_TolC"
rm(metal_efflux_RND_TolC)
```

There are also "heavy metal translocating P-type ATPase" BLAST annotations, which often correspond to Zn2+/Cd2+-exporting ATPase (EC:7.2.2.12 7.2.2.21) annotation from kofamscan.
Link here for more info on these: https://biocyc.org/gene?orgid=META&id=YHHO-MONOMER.
Looks like it's used for Zn homostasis and resistence to lead and cadmium, and has low activity with other divalent metals.
The "heavy metal translocating P-type ATPase" BLAST annotations encompass the kofamscan "Zn2+/Cd2+-exporting ATPase" annotations, so we'll use the former to mark them:

```{r metal_translocating_P_ATPase, message = FALSE}
metal_translocating_P_ATPase <- unique(annotation_data %>%
                                         filter(grepl("heavy metal translocating P-type ATPase", BLAST_gene_type, ignore.case = TRUE)) %>%
                                         select(seqID) %>% unlist(use.names = FALSE))
hgcA_gff$gene_info[which(hgcA_gff$seqID %in% metal_translocating_P_ATPase)] <- "metal_translocating_P_ATPase"
```

In the Actinobacteria gene neighborhood, there is a tellurite resistance protein (TerC).
Review paper: https://www.frontiersin.org/articles/10.3389/fmicb.2021.681175/full

```{r TerC, message = FALSE}
terC <- unique(annotation_data %>%
                 filter(grepl("TerC family protein", BLAST_gene_type, ignore.case = TRUE)) %>%
                 select(seqID) %>% unlist(use.names = FALSE))
hgcA_gff$gene_info[which(hgcA_gff$seqID %in% terC)] <- "terC"
rm(terC)
```



# Save out hgcA gene neighborhood data

```{r save_out_data}
test <- as.Date('2024-10-03')
write.csv(hgcA_gff,
          file = "dataEdited/bin_based_analyses/hgcA_geneNeighborhood_annotated.csv",
          quote = FALSE,
          row.names = FALSE)
```


# Label

For now, we'll stick with looking at the effect of the *arsR* sequence, and thus will manually annotate the genes based on it's presence.
First we'll look at the gene immediately preceeding *hgcA* and see if it's annotated as *arsR*.
This is a bit tricky since my gene neighborhood script flips all the scaffolds so the center genes to be in the positive.
It always starts at 5000 though, so let's take the highest gene that starts before that on each scaffold

```{r arsR_annotation_identify_preceeding_genes}
preceeding_genes <- hgcA_gff %>%
  filter(gene_start < 5000) %>%
  group_by(scaffoldID) %>%
  summarise(gene_start = max(gene_start)) %>%
  mutate(gene_preceeding_hgcA = "yes") %>%
  right_join(hgcA_gff) %>%
  filter(gene_preceeding_hgcA == "yes")
```

Next, we'll look at just the *hgcA* genes, and add a column with notes on what the preceeding gene is.


```{r assign_regulation_status }
regulation_notes <- hgcA_gff %>%
  filter(gene_info == "hgcA") %>%
  select(cluster_ID, taxonomic_assignment, metabolic_assignment, scaffoldID) %>%
  mutate(regulation_notes = NA)

regulation_notes$regulation_notes[which(!(regulation_notes$scaffoldID %in% preceeding_genes$scaffoldID))] <- "no_gene"
regulation_notes$regulation_notes[which(regulation_notes$scaffoldID %in% (preceeding_genes %>%
                                                                              filter(gene_info == "transc_reg_arsR") %>%
                                                                              select(scaffoldID) %>% unlist(use.names = FALSE)))] <- "arsR"
regulation_notes$regulation_notes[which(regulation_notes$scaffoldID %in% (preceeding_genes %>%
                                                                              filter(gene_info == "transc_reg_marR") %>%
                                                                              select(scaffoldID) %>% unlist(use.names = FALSE)))] <- "marR"
# Some have a arsenate reductase, arsenite transporter, or terC. We'll call those none, as in, no transcriptional regulation.
regulation_notes$regulation_notes[which(regulation_notes$scaffoldID %in% (preceeding_genes %>%
                                                                              filter(gene_info == "ACR3_arsenite_efflux") %>%
                                                                              select(scaffoldID) %>% unlist(use.names = FALSE)))] <- "none"
regulation_notes$regulation_notes[which(regulation_notes$scaffoldID %in% (preceeding_genes %>%
                                                                              filter(gene_info == "ArsC_arsenate_reductase") %>%
                                                                              select(scaffoldID) %>% unlist(use.names = FALSE)))] <- "none"
regulation_notes$regulation_notes[which(regulation_notes$scaffoldID %in% (preceeding_genes %>%
                                                                              filter(is.na(gene_info)) %>%
                                                                              select(scaffoldID) %>% unlist(use.names = FALSE)))] <- "none"
regulation_notes$regulation_notes[which(regulation_notes$scaffoldID %in% (preceeding_genes %>%
                                                                              filter(gene_info == "terC") %>%
                                                                              select(scaffoldID) %>% unlist(use.names = FALSE)))] <- "none"

```

Here I want to specifically target which clusters are going to have *arsR*.
The other clear transcriptional regulator preceeding *hgcA* is *marR*.
For the *hgcA* genes that don't have an ORF upstream (meaning, the scaffold is truncated), we'll label them "unknown".
Three groups: arsR, marR, none, and unknown.
We'll combine these into a named vector that we can read in to other analytical scripts.

```{r regulation_vector}
clusters_with_arsR <- regulation_notes %>%
  filter(regulation_notes == "arsR") %>%
  select(cluster_ID) %>% unlist(use.names = FALSE) %>% unique()

clusters_with_marR <- regulation_notes %>%
  filter(regulation_notes == "marR") %>%
  select(cluster_ID) %>% unlist(use.names = FALSE) %>% unique()
# Only keep ones that aren't annotated as arsR already
clusters_with_marR <- clusters_with_marR[!(clusters_with_marR %in% clusters_with_arsR)]

clusters_with_no_reg <- regulation_notes %>%
  filter(regulation_notes == "none") %>%
  select(cluster_ID) %>% unlist(use.names = FALSE) %>% unique()
# Only keep ones that aren't annotated elsewhere
clusters_with_no_reg <- clusters_with_no_reg[!(clusters_with_no_reg %in% c(clusters_with_arsR, clusters_with_marR))]

clusters_with_unknown_reg <- regulation_notes %>%
  filter(regulation_notes == "no_gene") %>%
  select(cluster_ID) %>% unlist(use.names = FALSE) %>% unique()
# Only keep ones that aren't annotated elsewhere
clusters_with_unknown_reg <- clusters_with_unknown_reg[!(clusters_with_unknown_reg %in% c(clusters_with_arsR, clusters_with_marR, clusters_with_no_reg))]

list_of_vectors <- list(arsR = clusters_with_arsR, marR = clusters_with_marR,
                        no_reg = clusters_with_no_reg, unknown = clusters_with_unknown_reg)
regulation_vector <- vector()
for (annotation_vector_index in 1:length(list_of_vectors)) {
  name_to_use <- names(list_of_vectors)[annotation_vector_index]
  vector_to_use <- list_of_vectors[[annotation_vector_index]]
  temp_vector <- rep(name_to_use, length(vector_to_use))
  names(temp_vector) <- vector_to_use
  regulation_vector <- c(regulation_vector, temp_vector)
}
rm(list_of_vectors, clusters_with_arsR, clusters_with_marR, clusters_with_no_reg, clusters_with_unknown_reg)

saveRDS(object = regulation_vector,
        file = "working_directory/transcriptional_regulation_vector.rds")
```


# Check on cluster of Desulfo

These three genes were annotated as Desulfobacterota because one of them was binned (BLI_hgcA_HMS_0012):

BLI21_coassembly_000000025634_9
BLI21_coassembly_000000086306_5
BLI21_coassembly_000000502164_1

However, phylogenetically they were closer to PVC, so I wanted to check them here.
First, I pulled in the bin data to refresh my memory on this organism.

```{r bin_data_for_HMS_0012}
bin_data <- readRDS(file = 'dataEdited/bin_based_analyses/bin_data_aggregate_2_metabolism.rds') %>%
  filter(hms_id == "BLI_hgcA_HMS_0012")
```

Oops, I had originally called this bin Desulfobacterales.
Not the case.
It's in the Desulfobacterota phylum, down to the Desulfomonilaceae family.
I had thought this family was clustered in with Firmicutes/Clostridia, including *Desulfomonile tiedjei* for some reason.
The two wrongs somewhat cancel out though, as this cluster does seem to be a SRB.
Provides more confidence that the annotation is correct as well, given that these aren't necessarily closely related to the other SRBs. 

```{r check_SRB_hgcA }
hgcA_scaffolds_to_check <- c("BLI21_coassembly_000000025634",
                             "BLI21_coassembly_000000086306",
                             "BLI21_coassembly_000000502164")
clusters_to_check <- annotation_data %>%
  filter(scaffoldID %in% hgcA_scaffolds_to_check) %>%
  select(cluster_ID) %>% unlist(use.names = FALSE) %>%
  unique()

check_data <- annotation_data %>%
  filter(cluster_ID %in% clusters_to_check)
unique(check_data$scaffoldID)
```

Many of the genes here are annotated as *Desulfomonile*, *Desulfobacter*, so I think our taxonomic affiliation is correct.
Interestingly, several of the *hgcA* genes are annotated as *Anaerocolumna jejuensis*, which is a Clostridia.
Perhaps there was a transfer of the *hgcA* gene.
Either way, safe to continue with these as SRBs, and I fixed the annotation for these *hgcA* seqs in `hgcA_data_processing.Rmd`.


# Save out gene neighborhood information

I'm going to want to look at the *hgcA* gene neighborhood of specific bins, so let's add that info here. 
```{r save_gene_neighborhood_info}
write.csv(hgcA_gff,
          file = "dataFinal/hgcA_GNs.csv",
          quote = FALSE,
          row.names = FALSE)
```
