---
title: "hgcA+ bins - data aggregation"
author: "Benjamin D. Peterson"
date: "2023-11-17"
output: html_document
---

```{r setup, include=FALSE}
rm(list = ls())
#setwd("/Users/benjaminpeterson/Documents/research/BLiMMP")
# new.wd <- "/Users/benjaminpeterson/Documents/research/BLiMMP"
library(ggpubr)
library(lubridate)
library(readxl)
library(tidyverse)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(echo = TRUE)
new.wd <- gsub("\\/code/binBasedAnalysis$", "", getwd())
knitr::opts_knit$set(root.dir = new.wd)
rm(new.wd)
```

First I read in the information linking the bin to the assembly it was generated from and the ID of the *hgcA* gene in it. One of the bins had two *hgcA* genes, hence the summarize step.

```{r read_bin_assembly_hgcA_info, message = FALSE, fig.height = 6, fig.width = 6}
hgcA_bin_data <- read.table("dataEdited/bin_based_analyses/final_bin_data/bin_to_assembly.tsv",
                            col.names = c("bin_id", "assembly_id")) %>%
  right_join(read.table("dataEdited/bin_based_analyses/final_bin_data/hgcA_to_bin.tsv",
                            col.names = c("hgca_id", "bin_id")) %>%
                          group_by(bin_id) %>% 
                          summarize(hgca_id = paste0(hgca_id, collapse = ",")))
```

Also in this, I'll want to add the *hgcA* cluster ID, as this is how I assigned transcriptional regulator presence to *hgcA* genes.

```{r hgcA_cluster_info, message = FALSE, fig.height = 6, fig.width = 6}
hgcA_bin_data <- hgcA_bin_data %>%
   mutate(scaffold_id = paste(hgca_id %>% strsplit("_") %>% sapply("[", 1),
                              hgca_id %>% strsplit("_") %>% sapply("[", 2),
                              hgca_id %>% strsplit("_") %>% sapply("[", 3),
                              sep = '_')) %>%
  left_join(read.csv("dataFinal/hgcA_data.csv") %>%
              # filter(verified_hgcA) %>%
              select(seqID, cluster_ID) %>%
              mutate(scaffold_id = paste(seqID %>% strsplit("_") %>% sapply("[", 1),
                                         seqID %>% strsplit("_") %>% sapply("[", 2),
                                         seqID %>% strsplit("_") %>% sapply("[", 3),
                                         sep = '_')) %>%
              select(-seqID)) %>%
  select(-scaffold_id)
```

Now I'll add the bin cluster information as generated by Sarah Stevens's ANI calculator workflow.

```{r grouping_bins_by_ANI}
#### Calculate HMSs ####
ANI_values <- read.table(file = "dataEdited/bin_based_analyses/final_bin_data/ANI_comparison.txt",
                         sep = "\t",
                         header = FALSE,
                         col.names = c("GENOME1", "GENOME2", "ANI_1", "ANI_2", "AF_1", "AF_2"),
                         stringsAsFactors = FALSE) %>%
  mutate(GENOME1 = gsub(".fna", "", GENOME1),
         GENOME2 = gsub(".fna", "", GENOME2))

# Check plot of values
plot(y = ANI_values$ANI_1,
     x = ANI_values$AF_1,
     xlim = c(0.2, 1.0),
     ylim = c(70, 100),
     xlab = "Alignment fraction",
     ylab = "Average nucleotide identity",
     pch = 18)
points(y = ANI_values$ANI_2,
       x = ANI_values$AF_2,
       pch = 18)
abline(h = 97)
abline(v = 0.3)
```

Normally I would want to set the alignment fraction a little higher. However, we do have a mix of Das Tool vs. anvi'o generated bins, so there may be a wide range of what was included in the bin. Additionally, I'm keeping all of the bins and analyzing them as a cluster, so if there's a bin set with high ANI but low AF with different metabolic capacity, I can identify that and look into it further. Let's use a 0.95 ANI and 0.3 AF cutoff, as displayed in the graph above. Using this cutoff, I'll generate groups with the graph_from_data_frame function from the igraph package.

```{r set_cut_offs_generate_groups }
ANI_cutoff <- 97
cov_cutoff <- 0.5

# Generate edgelist
edgelist <- ANI_values %>%
  filter((ANI_1 > ANI_cutoff |
           ANI_2 > ANI_cutoff) &
           (AF_1 > cov_cutoff |
           AF_2 > cov_cutoff)) %>%
  select(GENOME1, GENOME2)
# Generate the graph from the edgelist
adjacency_graph <- igraph::graph_from_data_frame(edgelist)
# Store the bin and HMS name in a df.
HMS_ID <- data.frame(hms_id = paste("BLI_hgcA_HMS_",
                                    stringr::str_pad(igraph::clusters(adjacency_graph)$membership, 4, pad = "0"),
                                    sep = ""),
                     bin_id = names(igraph::clusters(adjacency_graph)$membership),
                     stringsAsFactors = FALSE)
```

Now I have a set of all the bins that belong to a HMS. Total of `r length(unique(HMS_ID$HMS_id))` HMSs, with a total of `r dim(HMS_ID)[1]` bins in those HMSs. That leaves us with `r dim(hgcA_bin_data)[1] - dim(HMS_ID)[1]` bins that are not in an HMS. I want to assign an HMS_id to those "loner" bins. We'll call them "BLI_hgcA_bin_XXX".

```{r check_out_loner_bins }
# Check out loner bins
lone_bins <- data.frame(bin_id = hgcA_bin_data$bin_id[!(hgcA_bin_data$bin_id %in% HMS_ID$bin_id)],
                        stringsAsFactors = FALSE) %>%
  mutate(hms_id = paste("BLI_hgcA_bin_", stringr::str_pad(row_number(), 4, pad = "0"),
                        sep = ""))

# Add loner bins to list
HMS_ID <- do.call("rbind",
                   list(HMS_ID, lone_bins))
```

Now I can add this HMS information into the main data tab.

```{r add_hms_data, message = FALSE}
hgcA_bin_data <- hgcA_bin_data %>%
  left_join(HMS_ID)
```

Next I can add the completeness and redundancy information from CheckM:

```{r checkM, message=FALSE}
checkm_data <- read.table("dataEdited/bin_based_analyses/final_bin_data/checkM_stats_clean.tsv",
                          sep = ',',
                          col.names = c("bin_id", "completeness", "contamination", "strain_het", "genome_length", "contig_count",
                                        "N50", "mean_contig_length", "longest_contig", "GC", "predicted_genes"))
hgcA_bin_data <- hgcA_bin_data %>%
  left_join(checkm_data)
```

Next I can add the taxonomy data from GTDB.

```{r add taxonomy }
taxonomy_data <- read.table("dataEdited/bin_based_analyses/final_bin_data/taxonomy_summary.txt",
                            sep = "\t",
                            header = TRUE,
                            col.names = c("bin_id", "gtdb_tax"),
                            stringsAsFactors = FALSE)
hgcA_bin_data <- hgcA_bin_data %>%
  left_join(taxonomy_data)
rm(taxonomy_data)
```

This is sufficient for our initial analyses. In other files, we'll add data on the abundance, transcription, gene neighborhoods of *hgcA*, and metabolic potential of the hgcA+ bins.

```{r save_out_data }
saveRDS(hgcA_bin_data,file = 'dataEdited/bin_based_analyses/bin_data_aggregate_1.rds')
```


```{r cleanup_round1, include = FALSE}
rm(adjacency_graph, ANI_values, edgelist, ANI_cutoff, cov_cutoff, lone_bins, HMS_ID, checkm_data, taxonomy_data)
```

